### DISCUSSION TITLE
What information is given by the descriptors FpMorganDensity1, FpMorganDensity2 and FpMorganDensity3?

### DISCUSSION THREAD

**Discussion by halexmmond** [rdkit/rdkit#6721] [Q&A]
Hi there,

I'm using rdkit to calculate molecular descriptors for a series of chemicals in order to predict acute toxicity values. I'm relatively new to computational chemistry and have been reading up on each of the available molecular descriptors. I can't seem to find information on a Morgan fingerprint density anywhere. I have looked into ECFPs and Morgan fingerprints and just want to be sure I understand the descriptors fully before using them in my project. Can anyone provide me with a brief description of what calculations are made to obtain these values?

Thanks in advance for any advice!

**Comment by dehaenw** [comment]
Hi,
this is simply the density (= count of on-bits in the fingerprint divided by the heavy atom count in the molecule) of the morgan fingerprint. this means this will be a count of the total amount of unique morgan-type atomic environments (up to radius 1,2,3) in your molecule normalized by the heavy atom count. With some hand-waving, you can consider this as an atom+substructure count with a downwards penalty for any redundancies like atom types that occur more than once, and substructures that occur more than once, with a normalization for atom count. As an example, consider these two C6H12 isomers:
```python
from rdkit import Chem
from rdkit.Chem import Descriptors

mol1 = Chem.MolFromSmiles("C1CCCCC1")
mol2 = Chem.MolFromSmiles("CC1CC1CC")
for m in [mol1,mol2]:
    display(m)
    print(Descriptors.FpDensityMorgan3(m))
```

giving 0.666... for cyclohexane (mol1), but giving 2.166.. for its less symmetrical isomer 1-Ethyl-2-methylcyclopropane (mol2). Remark that what is considered a "unique" atom is determined by the atom invariants used for morgan fingerprints, meaning different atom type, degree, ring membership, etc correspond to a different atom, and what is considered a "unique" fragment is based on rdkit's implementation of the Morgan algorithm.

**Reply by halexmmond** [reply]
> Hi,
> this is simply the density (= count of on-bits in the fingerprint divided by the heavy atom count in the molecule) of the morgan fingerprint. this means this will be a count of the total amount of unique morgan-type atomic environments (up to radius 1,2,3) in your molecule normalized by the heavy atom count. With some hand-waving, you can consider this as an atom+substructure count with a downwards penalty for any redundancies like atom types that occur more than once, and substructures that occur more than once, with a normalization for atom count. As an example, consider these two C6H12 isomers:
> ```python
> from rdkit import Chem
> from rdkit.Chem import Descriptors
> 
> mol1 = Chem.MolFromSmiles("C1CCCCC1")
> mol2 = Chem.MolFromSmiles("CC1CC1CC")
> for m in [mol1,mol2]:
>     display(m)
>     print(Descriptors.FpDensityMorgan3(m))
> ```
> 
> giving 0.666... for cyclohexane (mol1), but giving 2.166.. for its less symmetrical isomer 1-Ethyl-2-methylcyclopropane (mol2). Remark that what is considered a "unique" atom is determined by the atom invariants used for morgan fingerprints, meaning different atom type, degree, ring membership, etc correspond to a different atom, and what is considered a "unique" fragment is based on rdkit's implementation of the Morgan algorithm.
> 

Thanks so much for your quick and detailed response!
